### PR TITLE
pybind/mgr: add typing annotation

### PR DESCRIPTION
**Pull request by tchaikov**
Fixes: https://tracker.ceph.com/issues/48055

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by tchaikov** [comment]
jenkins test make check

File: src/mypy.ini [hunk 1]
```diff
@@ -27,13 +27,18 @@ disallow_untyped_defs = True
...
 [mypy-devicehealth.*]
 disallow_untyped_defs = True
 
+[mypy-hello.*]
+disallow_untyped_defs = True
+
 [mypy-iostat.*]
 disallow_untyped_defs = True
 
->>>>>>> 514d3b6785... mgr/rook: disallow_untyped_defs = True
```
Comment (tchaikov) [inline src/mypy.ini:36]: @sebastian-philipp i am sorry for introducing this change. this change is included by #39198.

Comment (sebastian-philipp) [reply src/mypy.ini:36]: no worries. On the contrary, thanks for merging for lots of cephadm PRs yesterday

File: src/pybind/mgr/telemetry/module.py [hunk 2]
```diff
@@ -187,81 +200,95 @@ def config_notify(self):
...
             metadata[key] = defaultdict(int)
 
         for osd in osd_map['osds']:
-            res = self.get_metadata('osd', str(osd['osd'])).items()
+            res = self.get_metadata('osd', str(osd['osd']))
             if res is None:
                 self.log.debug('Could not get metadata for osd.%s' % str(osd['osd']))
                 continue
-            for k, v in res:
+            for k, v in res.items():
```
Comment (sebastian-philipp) [inline src/pybind/mgr/telemetry/module.py:197]: This has never worked as intended.

File: src/pybind/mgr/telemetry/module.py [hunk 3]
```diff
@@ -187,81 +200,95 @@ def config_notify(self):
...
+                                                   bool,
+                                                   List[int],
+                                                   Dict[str, int],
+                                                   Dict[int, int]]]:
         osdmap = self.get_osdmap()
         crush_raw = osdmap.get_crush()
         crush = crush_raw.dump()
 
-        def inc(d, k):
+        T = TypeVar('T', int, str)
```
Comment (sebastian-philipp) [inline src/pybind/mgr/telemetry/module.py:236]: I'd be :+1: if we could the generic name  `T` only for generic type vars. Maybe something like `IntStrT`?

File: src/pybind/mgr/alerts/module.py [hunk 4]
```diff
@@ -3,73 +3,65 @@
...
-from mgr_module import CLIReadCommand, MgrModule, HandleCommandResult
+from mgr_module import CLIReadCommand, HandleCommandResult, MgrModule, Option
 from threading import Event
 from typing import Any, Optional, Dict, List, TYPE_CHECKING, Union
 import json
 import smtplib
 
+
 class Alerts(MgrModule):
     # ´# type: ignore` due to the introduction of the Option type.
```
Comment (sebastian-philipp) [inline src/pybind/mgr/alerts/module.py:14]: you can remove this line now

File: src/pybind/mgr/crash/module.py [hunk 5]
```diff
@@ -222,22 +239,41 @@ def do_ls(self, cmd, inbuf):
...
-    def do_prune(self, cmd, inbuf):
-        keep = cmd['keep']
-        try:
-            keep = int(keep)
-        except ValueError:
-            return errno.EINVAL, '', 'keep argument must be integer'
-
+    @CLICommand('crash prune')
+    @with_crashes
+    def do_prune(self, keep: int) -> Tuple[int, str, str]:
```
Comment (sebastian-philipp) [inline src/pybind/mgr/crash/module.py:289]: I'm not sure if this works! `@with_crashes` removes the annotations for the parameters and `@CLICommand` takes the annotations for parameters to compute the arguments for this MON command.

File: src/pybind/mgr/crash/module.py [hunk 6]
```diff
@@ -1,20 +1,34 @@
...
+from typing import Any, Dict, Optional, Tuple
 
 DATEFMT = '%Y-%m-%dT%H:%M:%S.%f'
 OLD_DATEFMT = '%Y-%m-%d %H:%M:%S.%f'
 
 MAX_WAIT = 600
 MIN_WAIT = 60
 
+
+def with_crashes(func: HandlerFuncType) -> HandlerFuncType:
```
Comment (sebastian-philipp) [inline src/pybind/mgr/crash/module.py:24]: I would use 

```python
FuncT = TypeVar('FuncT', bound=Callable)
```

here to keep the parameters annotated.

**Review by sebastian-philipp — CHANGES_REQUESTED** [review]
awesome work!

**Comment by sebastian-philipp** [comment]
mypy:

```
crash/__init__.py:1: note: In module imported here:
crash/module.py: note: In member "serve" of class "Module":
crash/module.py:73: error: Argument 1 to "_prune" of "Module" has incompatible type "float"; expected "int"
crash/module.py: note: In member "handle_bcommand" of class "Module":
crash/module.py:121: error: Function is missing a return type annotation
crash/module.py:121: error: Function is missing a type annotation for one or more arguments
crash/module.py: note: In member "calc_sig" of class "Module":
crash/module.py:189: error: Function is missing a type annotation for one or more arguments
crash/module.py: note: In member "ls" of class "Module":
crash/module.py:233: error: Function is missing a return type annotation
```

File: src/pybind/mgr/crash/module.py (hunk 5 above)
Comment (tchaikov) [reply src/pybind/mgr/crash/module.py:289]: yeah, that's why i added this before returning the wrapper.
```py
wrapper.__signature__ = inspect.signature(func)
```

Comment (sebastian-philipp) [reply src/pybind/mgr/crash/module.py:289]: And this is needed, independent of what annotations we use for `with_crashes`.

File: src/pybind/mgr/hello/module.py [hunk 7]
```diff
@@ -93,34 +76,43 @@ def config_notify(self):
...
+        """
+        Say hello
+        """
+        out = ''
+        if person_name is None:
+            out = 'Hello, ' + cast(str, self.get_module_option('place'))
+        else:
+            out = f'Hello, {person_name}'
+        if self.get_module_option('emphatic'):
+            out += '!'
```
Comment (epuertat) [inline src/pybind/mgr/hello/module.py:91]: Why not promoting adoption of f-strings vs. concat?
```suggestion
        out = f"Hello, {person_name or self.get_module_option('place')}{'!' if emphatic else ''}"
```

File: src/pybind/mgr/prometheus/module.py [hunk 8]
```diff
@@ -516,7 +517,8 @@ def _get_value(message, delim=' ', word_pos=0):
                         v, err = _get_value(message)
 
                     if err:
-                        self.log.error("healthcheck {} message format is incompatible and has been dropped".format(check.name))
+                        self.log.error(
+                            "healthcheck {} message format is incompatible and has been dropped".format(check.name))
```
Comment (epuertat) [inline src/pybind/mgr/prometheus/module.py:521]: Nit: `logging` supports [lazy interpolation with `str.format()`](https://docs.python.org/3/library/logging.html#logging.Formatter)

**Review by epuertat — APPROVED** [review]
Awesome clean-up (and some bugfixing), @tchaikov!

File: src/pybind/mgr/alerts/module.py (hunk 4 above)
Comment (tchaikov) [reply src/pybind/mgr/alerts/module.py]: removed.

File: src/pybind/mgr/crash/module.py (hunk 6 above)
Comment (tchaikov) [reply src/pybind/mgr/crash/module.py]: incorporated.

File: src/pybind/mgr/hello/module.py (hunk 7 above)
Comment (tchaikov) [reply src/pybind/mgr/hello/module.py]: thanks. add a separate commit to promote f-strings.

File: src/pybind/mgr/prometheus/module.py (hunk 8 above)
Comment (tchaikov) [reply src/pybind/mgr/prometheus/module.py]: thanks for your suggestion, updated accordingly.

File: src/pybind/mgr/telemetry/module.py (hunk 3 above)
Comment (tchaikov) [reply src/pybind/mgr/telemetry/module.py]: renamed to `BucketKeyT`.

**Comment by tchaikov** [comment]
@sebastian-philipp and @epuertat thanks for your thorough reviews. updated and repushed.

File: src/pybind/mgr/hello/module.py (hunk 7 above)
Comment (sebastian-philipp) [reply src/pybind/mgr/hello/module.py]: I somewhat don't like `{... if ... else ''}` constructs within f-strings. I'm ok with them, but they feel a bit clunky

Comment (tchaikov) [reply src/pybind/mgr/hello/module.py]: neither do i, i ended up with
```py
        fin = '!' if self.get_module_option('emphatic') else ''
        return HandleCommandResult(stdout=f'Hello, {who}{fin}')
```

**Review by sebastian-philipp — APPROVED** [review]

**Review by yaarith — APPROVED** [review]
Great work, @tchaikov, looks good to me!

File: src/pybind/mgr/telemetry/module.py (hunk 2 above)
Comment (yaarith) [reply src/pybind/mgr/telemetry/module.py:197]: it was partly working before the change, except for the case when `res` was `None`...
thanks for fixing this, @tchaikov!

**Comment by yaarith** [comment]
@dmick @liewegas might want to take a look as well?

File: src/pybind/mgr/dashboard/controllers/telemetry.py [hunk 9]
```diff
@@ -235,6 +235,6 @@ def singleton_set(self, enable=True, license_name=None):
                         'Sharing - Version 1.0 (https://cdla.io/sharing-1-0/). To enable, add '
                         '{"license": "sharing-1-0"} to the request payload.'
                 )
-            mgr.remote('telemetry', 'on')
+            mgr.remote('telemetry', 'on', license_name)
```
[the remark this answers is not in the document]
Comment (tchaikov) [reply src/pybind/mgr/dashboard/controllers/telemetry.py:238]: this is not a fix. it's just to accommodate the change on the telemetry side where `Module.on()` now requires a `license_name` parameter. before this change, `Module.on()` was just a helper method. after this change, `Module.on()` now serves as a handler of the `telemetry on` command.

File: src/pybind/mgr/telemetry/module.py [hunk 10]
```diff
@@ -587,7 +549,7 @@ def compile_report(self, channels=[]):
                                                  'mds.root_rbytes')
                         rsnaps = self.get_latest('mds', mds['name'],
                                                  'mds.root_rsnaps')
-                report['fs']['filesystems'].append({
+                report['fs']['filesystems'].append({  # type: ignore
```
[the remark this answers is not in the document]
Comment (tchaikov) [reply src/pybind/mgr/telemetry/module.py:552]: `report['fs']` references a dict holding heterogeneous data structures including `int` (count), an other dict (feature_flags), and a list of filesystem. but mypy cannot deduce the type of value by the value of key. without ignoring the type here, i'd have to cast `report['fs']['filesystems']` to a list or create a temporary dict and then assign it to `report['fs']['filesystems']`.

**Comment by tchaikov** [comment]
changelog

- decorate all crash commands with `@with_crashes`. some of them were missing it in the previous revision.

File: src/pybind/mgr/tox.ini [hunk 11]
```diff
@@ -57,8 +57,10 @@ commands =
     mypy --config-file=../../mypy.ini \
            -m alerts \
            -m cephadm \
+           -m crash \
            -m dashboard \
            -m devicehealth \
+           -m hello \
```
Comment (epuertat) [inline src/pybind/mgr/tox.ini:63]: Missing? The `fix` environment included `insights`, but not the `mypy`.
```suggestion
           -m hello \
           -m insights \
```

File: src/pybind/mgr/tox.ini [hunk 12]
```diff
@@ -84,7 +87,17 @@ commands = {[testenv]commands}
...
+modules =
+    cephadm
+    crash
+    devicehealth
+    insights
+    iostat
+    orchestrator
+    prometheus
+    status
+    telemetry
```
Comment (epuertat) [inline src/pybind/mgr/tox.ini:99]: Comparing this to the `mypy` env I find the following missing:
```suggestion
modules =
    alerts
    cephadm
    crash
    dashboard
    devicehealth
    insights
    hello
    iostat
    mds_autoscaler
    mgr_module
    mgr_util
    mirroring
    orchestrator
    progress
    prometheus
    rook
    status
    telemetry
    volumes
    zabbix
```

File: src/pybind/mgr/tox.ini (hunk 11 above)
Comment (tchaikov) [reply src/pybind/mgr/tox.ini:63]: i haven't added the type annotations to the `insights` module yet.

File: src/pybind/mgr/tox.ini (hunk 12 above)
Comment (tchaikov) [reply src/pybind/mgr/tox.ini:99]: the PEP8-clean list is not in sync with that of mypy-clean list yet. they will.

**Comment by tchaikov** [comment]
changelog

- fixed `crash stat` command. use `bins.append(...)` instead of `bins[i]`. as we don't reuse `bins` for `AGE_IN_DAYS` in this PR anymore.

**Comment by tchaikov** [comment]
```
[ RUN      ] TestLibRBD.TestPendingAio
using new format!
/home/jenkins-build/build/workspace/ceph-pull-requests/src/test/run-rbd-unit-tests.sh: line 20: 24606 Segmentation fault      (core dumped) RBD_FEATURES=$i unittest_librbd
```

**Comment by tchaikov** [comment]
jenkins test make check

**Comment by tchaikov** [comment]
jenkins test make check

**Comment by tchaikov** [comment]
- https://pulpito.ceph.com/kchai-2021-02-02_15:01:12-rados-wip-kefu-testing-2021-02-02-1939-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2021-02-03_08:10:00-rados-wip-kefu-testing-2021-02-03-1357-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2021-02-03_14:17:43-rados-wip-kefu-testing-2021-02-03-2051-distro-basic-smithi/

**Pull request merged**